File: app/src/main/java/com/sample/app/fragment_with_dialog/FragmentWithDialog.java
package com.sample.app.fragment_with_dialog;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.sample.app.R;
import com.sample.app.fragment_with_dialog.dialogs.DialogWithLayout;

public class FragmentWithDialog extends Fragment implements View.OnClickListener {
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView      = inflater.inflate(R.layout.fragment_with_dialog, container, false);

        Button button1 = (Button) rootView.findViewById(R.id.button_alert_dialog);
        Button button2 = (Button) rootView.findViewById(R.id.button_layout_dialog);

        button1.setOnClickListener(this);
        button2.setOnClickListener(this);

        return rootView;
    }

    @Override
    public void onClick(View v) {
        switch( v.getId() ) {
            case R.id.button_alert_dialog: {
                createAlertDialog();
                break;
            }
            case R.id.button_layout_dialog: {
                startDialog();
                break;
            }
        }
    }

    private void createAlertDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle("Title");
        builder.setMessage("Message");
        builder.setPositiveButton("Ok",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        // action when ok pressed
                    }
                });
        builder.setNegativeButton("Cancel",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        // action when cancel pressed
                    }
                });
        AlertDialog alertDialog = builder.show();
        alertDialog.setCanceledOnTouchOutside(false);
    }

    private void startDialog() {
        DialogWithLayout dialog = new DialogWithLayout();
        dialog.setTargetFragment(FragmentWithDialog.this, 0);
        dialog.show(getChildFragmentManager(), "Custom Dialog");
    }

}

File: app/src/main/java/com/sample/app/fragment_with_dialog/FragmentWithDialog.kt
package com.sample.app.fragment_with_dialog

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import com.sample.app.R
import com.sample.app.fragment_with_dialog.dialogs.DialogWithLayout

class FragmentWithDialog : Fragment(), View.OnClickListener {
    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val rootView = inflater!!.inflate(R.layout.fragment_with_dialog, container, false)

        val button1 = rootView.findViewById(R.id.button_alert_dialog) as Button
        val button2 = rootView.findViewById(R.id.button_layout_dialog) as Button

        button1.setOnClickListener(this)
        button2.setOnClickListener(this)

        return rootView
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.button_alert_dialog -> {
                createAlertDialog()
            }
            R.id.button_layout_dialog -> {
                startDialog()
            }
        }
    }

    private fun createAlertDialog() {
        val builder = AlertDialog.Builder(context)
        builder.setTitle("Title")
        builder.setMessage("Message")
        builder.setPositiveButton("Ok"
        ) { dialog, which ->
            // action when ok pressed
        }
        builder.setNegativeButton("Cancel"
        ) { dialog, which ->
            // action when cancel pressed
        }
        val alertDialog = builder.show()
        alertDialog.setCanceledOnTouchOutside(false)
    }

    private fun startDialog() {
        val dialog = DialogWithLayout()
        dialog.setTargetFragment(this@FragmentWithDialog, 0)
        dialog.show(childFragmentManager, "Custom Dialog")
    }

}
